app/controllers/voucher_adjustments_controller.rb

```diff
@@ -1,32 +1,66 @@
 # frozen_string_literal: true
 
 class VoucherAdjustmentsController < BaseController
-  include CablecarResponses
+  before_action :set_order
 
-  def destroy
-    @order = current_order
+  def create
+    if add_voucher
+      render_voucher_section
+    elsif @order.errors.present?
+      render_error
+    end
+  end
 
+  def destroy
     @order.voucher_adjustments.find_by(id: params[:id])&.destroy
 
-    respond_to do |format|
-      format.cable_ready { render_voucher_section }
-      format.html { redirect_to checkout_step_path(:payment) }
-    end
+    render_voucher_section
   end
 
   private
 
-  # Using the power of cable_car we replace only the #voucher_section instead of reloading the page
+  def set_order
+    @order = current_order
+  end
+
+  def add_voucher
+    if params[:voucher_code].blank?
+      @order.errors.add(:voucher, I18n.t('split_checkout.errors.voucher_not_found'))
+      return false
+    end
+
+    voucher = Voucher.find_by(code: params[:voucher_code], enterprise: @order.distributor)
+
+    if voucher.nil?
+      @order.errors.add(:voucher, I18n.t('split_checkout.errors.voucher_not_found'))
+      return false
+    end
+
+    adjustment = voucher.create_adjustment(voucher.code, @order)
+
+    if !adjustment.valid?
+      @order.errors.add(:voucher, I18n.t('split_checkout.errors.add_voucher_error'))
+      adjustment.errors.each { |error| @order.errors.import(error) }
+      return false
+    end
+
+    true
+  end
+
   def render_voucher_section
-    render(
-      status: :ok,
-      cable_ready: cable_car.replace(
-        "#voucher-section",
-        partial(
-          "split_checkout/voucher_section",
-          locals: { order: @order, voucher_adjustment: @order.voucher_adjustments.first }
-        )
+    render cable_ready: cable_car.replace(
+      selector: "#voucher-section",
+      html: render_to_string(
```

Comment:
To avoid a double render error. We're just writing a little bit of HTML to a variable here, but in controllers Rails will blow up if you call the `render` method twice.